mono/metadata/boehm-gc.c

```diff
@@ -502,11 +514,26 @@ on_gc_heap_resize (size_t new_size)
 	mono_profiler_gc_heap_resize (new_size);
 }
 
+typedef struct {
+	char *start;
+	char *end;
+} RootData;
+
+static gpointer
+register_root (gpointer arg)
+{
+	RootData* root_data = arg;
+	g_hash_table_insert (roots, root_data->start, root_data->end);
+	return NULL;
+}
+
 int
 mono_gc_register_root (char *start, size_t size, void *descr, MonoGCRootSource source, const char *msg)
 {
-	/* for some strange reason, they want one extra byte on the end */
-	GC_add_roots (start, start + size + 1);
+	RootData root_data;
+	root_data.start = start;
+	root_data.end = start + size + 1;
```

Comment:
Should preserve the comment here. I believe the “strange reason”, which isn’t so strange, is that one-past-the-end is a valid pointer to an object in C (and C++), which should cause a value to be conservatively retained.